### PR TITLE
Add hazards dataset taxonomy relabeling using several teacher models

### PR DESCRIPTION
**Pull request by tsunamit**
**tldr**: **Successfully generated ~27,000 new MLC 0.5 hazard training samples - looking for feedback on this process.** These samples can be used to finetune evaluator base models on the MLC 0.5 taxonomy. 

We use teacher LLM models to relabel non-MLC taxonomy datasets (Beavertails 30k open source dataset), and we use the consensus across several teacher models to validate labels.

Details

- **Teacher models used**: Used Llama 3 70b instruct, and Mistral 8x22b instruct as teacher models to relabel 30k user prompt samples from Beavertails hazard dataset using MLC 0.5 taxonomy
- **The joined and cleaned dataset can be reviewed here** (please provide feedback!): https://docs.google.com/spreadsheets/d/1nlQv_z5PlNyT2fAZONdJmckMY5dPq769TO6mWg5dBpc/edit?usp=sharing
- **Can we trust the teacher model labels?** We plan to use the consensus between them as a proxy for ground truth. However, further work will be needed to confirm if teacher-model-consensus is an appropriate proxy for human ground truth alignment.
- **Required human review:** Any non-consensus or ambiguous samples go to a separate dataset for human review.
- **Code pipeline** we wrote should generalize to taxonomy updates for MLC 1.0

### CODE REVIEW

File: label/safety_labeler.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            return parsed_response_model
+
+        except json.JSONDecodeError as e:
+            print(f"Error decoding JSON response: {e}")
+            print(
+                f"Failed sample: {message}, message_type: {message_type}, response (should be json but is not): {response_json}"
+            )
+            return None
+
+    async def async_label(
```
Comment (tsunamit) [inline label/safety_labeler.py:125]: Here's where we do most of the teacher LLM labeling logic. I mostly use the async method so I can batch requests for higher token throughput

File: label/prompts.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}}
+"""
+
+LABEL_CHECKER_USER_PROMPT = """
+User message: {user_message}
+Is safe?: {is_safe}
+Assigned safety categories: {categories}
+"""
+
+LABELER_SYS_PROMPT = """
```
Comment (tsunamit) [inline label/prompts.py:28]: This is the primary teacher model labeling prompt

File: label/label_beavertails.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    else:
+        sample["labeled_is_safe"] = response.is_safe
+        sample["labeled_safety_categories"] = response.safety_categories
+        sample["label_reasoning"] = response.reasoning
+        sample["label_succeeded"] = True
+
+    return sample
+
+
+def main():
```
Comment (tsunamit) [inline label/label_beavertails.py:73]: Here's the main script to run the dataset hazard labeling

File: label/label_beavertails.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        batch_label_beaver, fn_kwargs={"labeler": labeler}, batched=True, batch_size=20
+    )
+    # Synchronous, 1 at a time. Good for testing
+    # result = working_dataset.map(lambda sample: label_beaver(sample, labeler))
+
+    # Save the updated dataset
+    current_datetime = datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+    model_used = labeler.labeler_model.name
+    dataset_size = f"{len(working_dataset)}_rows"
+    output_location = f"/Users/ryan/projects/mlc/modeltune/label/outputs/{model_used}_{dataset_size}_{current_datetime}.csv"
```
Comment (tybrs) [inline label/label_beavertails.py:121]: Hey, Ryan. Not huge priority but we should add an argparse argument for this dataset path.
